### PR TITLE
Generate ensemble predictions for the uncertainty estimation

### PR DESCRIPTION
**Pull request by yplitw**

### CODE REVIEW

File: cnn_framework/layers.py [hunk 1]
```diff
@@ -12,7 +12,9 @@
 class MoleculeConv(Layer):
     def __init__(self, units, inner_dim, depth=2, init_output='uniform',
                  activation_output='softmax', init_inner='identity',
-                 activation_inner='linear', scale_output=0.01, padding=False, **kwargs):
+                 activation_inner='linear', scale_output=0.01, padding=False, 
+                 dropout_rate_outer=0.0, dropout_rate_inner=0.0,
+                 padding_final_size=None, **kwargs):
```
Comment (cgrambow) [inline cnn_framework/layers.py:18]: Why is `padding_final_size` required in `MoleculeConv` considering that padding is removed prior to the convolution? What if `padding_final_size=None` and dropout is used? Will that work?

File: scripts/train_cnn.py [hunk 2]
```diff
@@ -167,7 +167,8 @@ def initialize_log(verbose, log_file_name):
                           out_dir=out_dir,
                           normalize=normalize)
     predictor.load_input(input_file)
-    predictor.load_parameters(param_path=weights_file)
+    if weights_file is not None:
+        predictor.load_parameters(weights_file)
```
Comment (cgrambow) [inline scripts/train_cnn.py]: I think this commit can be dropped.

File: cnn_framework/predictor.py [hunk 3]
```diff
@@ -419,14 +419,23 @@ def predict(self, molecule=None, molecule_tensor=None):
...
+                return y_pred[0][0], y_sigma[0][0]
         else:
+            y_pred = self.model.predict(molecule_tensor_array)        
+            if self.y_mean is not None and self.y_std is not None:
+                y_pred = y_pred * self.y_std + self.y_mean
+            if self.prediction_task == "Cp(cal/mol/K)":
+                return y_pred[0]
+            else:
+                return y_pred[0][0]   
             return y_pred[0][0]
```
Comment (cgrambow) [inline cnn_framework/predictor.py]: Remove this line; it's unreachable.

File: cnn_framework/predictor.py [hunk 4]
```diff
@@ -419,14 +419,23 @@ def predict(self, molecule=None, molecule_tensor=None):
...
-        if self.y_mean is not None and self.y_std is not None:
-            y_pred = y_pred * self.y_std + self.y_mean
-
-        if self.prediction_task == "Cp(cal/mol/K)":
-            return y_pred[0]
+        if sigma:
+            y_pred, y_sigma = self.model.predict(molecule_tensor_array, sigma=sigma)
+            if self.y_mean is not None and self.y_std is not None:
+                y_pred = y_pred * self.y_std + self.y_mean
+                y_sigma = y_sigma * self.y_std**0.5
```
Comment (cgrambow) [inline cnn_framework/predictor.py]: Does it make sense to do `y_std**0.5`? That wouldn't give the correct units.

File: cnn_framework/uncertainty.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    def reset_mask_id(self):
+        self.mask_id = 0        
+
+    def train_on_batch(self, x, y, **kwargs):   
+        seed = np.random.choice(self.seeds)      
+        self.gen_mask(seed)
+        loss = super(EnsembleModel,self).train_on_batch(x, y, **kwargs)
+        return loss
+
+
```
Comment (cgrambow) [inline cnn_framework/uncertainty.py]: Can you remove this whitespace line?

File: cnn_framework/cnn_model.py [hunk 6]
```diff
@@ -63,10 +62,9 @@ def build_model(embedding_size=512, attribute_vector_size=33, depth=5,
 
     if n_model is None:
         model = Model(input=inputs, output=y)
-    elif n_model != 0:
-        model = EnsembleModel(input=inputs, output=y, seeds=np.random.randint(0, 10e8, n_model).tolist())
     else:
-        model = EnsembleModel(input=inputs, output=y)
+        model = EnsembleModel(input=inputs, output=y, seeds=np.random.randint(0, 10e8, n_model).tolist())
```
Comment (cgrambow) [inline cnn_framework/cnn_model.py]: Are you using `10e8` to try to ensure that all your seeds are different? To generate a sample without replacement you could use `np.random.choice` with the `replace=False` option.

File: cnn_framework/cnn_model_test.py [hunk 7]
```diff
@@ -49,3 +50,30 @@ def test_save_model(self):
...
+
+        embedding_size = 300
+        attribute_vector_size = 10
+        hidden = 0
+        n_model = 10
+        dropout_rate_inner = 0.5
+        dropout_rate_outer = 0.5
+        dropout_rate_hidden = 0.5
+        dropout_rate_output = 0.5
+        padding_final_size = 50
```
Comment (cgrambow) [inline cnn_framework/cnn_model_test.py:64]: Either remove this or change the `50` below to `padding_final_size`.

File: cnn_framework/layers.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 from keras import activations, initializations
 from keras.engine.topology import Layer
 from keras.layers import merge
+import numpy as np
```
Comment (cgrambow) [inline cnn_framework/layers.py:10]: Can you squash this commit with the previous one where you add things to `layers.py`?

File: cnn_framework/predictor.py [hunk 9]
```diff
@@ -377,6 +379,13 @@ def kfcv_batch_train(self, folds, pretrained_weights=None,
         if self.save_tensors_dir is not None:
             if not self.keep_tensors:
                 shutil.rmtree(self.save_tensors_dir)
+    
+    def load_architecture(self, param_path=None):
```
Comment (cgrambow) [inline cnn_framework/predictor.py:381]: Great method! I wasn't aware of the `custom_objects` option.

File: cnn_framework/layers.py [hunk 10]
```diff
@@ -167,7 +167,11 @@ def get_config(self):
                   'init_inner': self.init_inner.__name__,
                   'activation_inner': self.activation_inner.__name__,
                   'activation_output': self.activation_output.__name__,
-                  'input_dim': self.input_dim,
-                  'depth': self.depth}
+                  'scale_output': self.scale_output,
```
Comment (cgrambow) [inline cnn_framework/layers.py:170]: Can you also squash this with the other commit?

File: cnn_framework/molecule_tensor.py [hunk 11]
```diff
@@ -175,7 +175,8 @@ def get_bond_attributes(molecule,
         for bonded_atom, bond in atom.bonds.iteritems():
             if not bonded_atom.isHydrogen() and (bond not in bond_attributes_dict):
 
-                attributes = [1]
+            #    attributes = [1]
```
Comment (cgrambow) [inline cnn_framework/molecule_tensor.py]: Could you clarify what these changes do (also those in the other commit that change this code)? Also will this mean that loading pretrained weights from before this PR will fail, i.e., do I have to retrain my large models?

File: cnn_framework/input_test.py [hunk 12]
```diff
@@ -26,6 +26,7 @@ def test_read_input_file(self):
         self.assertEqual(len(predictor_model.layers), 4)
```
Comment (cgrambow) [inline cnn_framework/input_test.py]: Pretty sure commit `f0ae4c` doesn't exist anymore. Probably best not to refer to the commit hash in the commit message. Perhaps just squash the two commits.

File: cnn_framework/data.py [hunk 13]
```diff
@@ -385,19 +385,20 @@ def prepare_full_train_data_from_file(datafile,
...
 
     # Get random number generator state so that we can shuffle multiple arrays
-    rng_state = np.random.get_state()
+    rng_state = rng.get_state()
 
     # Shuffle data in place
-    np.random.shuffle(X)
-    np.random.set_state(rng_state)
-    np.random.shuffle(y)
+    rng.shuffle(X)
```
Comment (cgrambow) [inline cnn_framework/data.py:385]: What is the purpose of the changes in this commit? As far as I can tell, the previous code has the same effect.

File: cnn_framework/data.py [hunk 14]
```diff
@@ -382,7 +382,7 @@ def prepare_full_train_data_from_file(datafile,
 
 
 @attr('helper')
-def split_test_from_train_and_val(X, y, extra_data=None, shuffle_seed=None, testing_ratio=0.1):
+def split_test_from_train_and_val(X, y, extra_data=None, shuffle_seed=0, testing_ratio=0.1):
```
Comment (cgrambow) [inline cnn_framework/data.py]: If this commit is necessary, can it be squashed with the previous one?

File: cnn_framework/molecule_tensor.py [hunk 15]
```diff
@@ -175,16 +175,18 @@ def get_bond_attributes(molecule,
         for bonded_atom, bond in atom.bonds.iteritems():
             if not bonded_atom.isHydrogen() and (bond not in bond_attributes_dict):
 
-#                attributes = [1]
-                attributes = []
+
```
Comment (cgrambow) [inline cnn_framework/molecule_tensor.py]: Remove this whitespace line.

File: cnn_framework/molecule_tensor.py [hunk 16]
```diff
@@ -175,16 +175,18 @@ def get_bond_attributes(molecule,
         for bonded_atom, bond in atom.bonds.iteritems():
             if not bonded_atom.isHydrogen() and (bond not in bond_attributes_dict):
 
-#                attributes = [1]
-                attributes = []
+            
                 if differentiate_bond_type:
+                    attributes = []
```
Comment (cgrambow) [inline cnn_framework/molecule_tensor.py]: See previous comment about this part of the code. Also, this commit can be squashed with the other one.

File: cnn_framework/uncertainty_test.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+import keras.backend as K
+import numpy as np
+from keras.layers import Input
+from uncertainty import RandomMask, EnsembleModel
+import time
+import unittest
+
+class TestRandomMask(unittest.TestCase):
+    def testRandomMask(self):
```
Comment (cgrambow) [inline cnn_framework/uncertainty_test.py]: Please change the names of these methods to use underscores and lowercase, e.g., `test_random_mask`.

File: cnn_framework/predictor_test.py [hunk 18]
```diff
@@ -49,18 +49,14 @@ def test_load_input(self):
         dense2 = self.predictor.model.layers[3]
 
         self.assertEqual(gfp.W_inner.shape.eval()[0], 4)
-        #self.assertEqual(gfp.W_inner.shape.eval()[1], 38)
-        #self.assertEqual(gfp.W_inner.shape.eval()[2], 38)
-        self.assertEqual(gfp.W_inner.shape.eval()[1], 37)
-        self.assertEqual(gfp.W_inner.shape.eval()[2], 37)
+        self.assertEqual(gfp.W_inner.shape.eval()[1], 38)
```
Comment (cgrambow) [inline cnn_framework/predictor_test.py]: Ah ok, I hadn't gotten to this commit yet in my review, so that probably answers my question about whether I can load already-trained weights. Also, please squash this with the commit where you introduce the changes reverted here (or just drop both of them).

File: cnn_framework/predictor_test.py [hunk 19]
```diff
@@ -104,14 +104,14 @@ def test_load_parameters(self):
 
         self.assertAlmostEqual(gfp.W_inner.eval()[0][0][0], 1.000, 3)
         self.assertAlmostEqual(gfp.b_inner.eval()[0][0][0], 0.000, 3)
-        self.assertAlmostEqual(gfp.W_output.eval()[0][0][0], 0.003, 3)
-        self.assertAlmostEqual(gfp.b_output.eval()[0][0][0], -0.843, 3)
+        self.assertAlmostEqual(gfp.W_output.eval()[0][0][0], 0.040, 3)
```
Comment (cgrambow) [inline cnn_framework/predictor_test.py]: Same comment here. Please squash and/or drop.

File: cnn_framework/molecule_tensor.py [hunk 20]
```diff
@@ -203,7 +203,7 @@ def get_bond_attributes(molecule,
 
     if not bond_attributes_dict:
         if differentiate_bond_type:
-            bond_attributes_dict['no_bond'] = np.array([0]*15, dtype=np.float32)
+            bond_attributes_dict['no_bond'] = np.array([0]*14, dtype=np.float32)
```
Comment (cgrambow) [inline cnn_framework/molecule_tensor.py:205]: This commit can also be squashed with one of the previous ones.

File: cnn_framework/test_data/ensemble_predictor/predictor_input.py [hunk 21]
```diff
@@ -0,0 +1,11 @@
+predictor_model(embedding_size=512, depth=3, scale_output=0.05,
```
Comment (cgrambow) [inline cnn_framework/test_data/ensemble_predictor/predictor_input.py:1]: The formatting in this file is weird. Can you fix it?

File: cnn_framework/uncertainty_test.py [hunk 22]
```diff
@@ -27,3 +27,23 @@ def testEnsembleModel(self):
         expected_std_output = np.array([[0.5,0,0.5],[0.5,0,0.5],[0,0.5,0.5]])
         self.assertTrue(np.array_equal(y_output[0],expected_y_output))
         self.assertTrue(np.array_equal(std_output[0],expected_std_output))
+    
+    def testEnsembleTrain(self):
```
Comment (cgrambow) [inline cnn_framework/uncertainty_test.py]: Squash this part with the other unit test commit for `uncertainty_test.py`. Maybe you could squash this entire commit with that one. Up to you.

File: cnn_framework/predictor.py [hunk 23]
```diff
@@ -414,7 +414,7 @@ def save_model(self, loss, inner_val_loss, mean_outer_val_loss, mean_test_loss,
             np.savez(fpath + '_mean_std.npz', mean=self.y_mean, std=self.y_std)
             logging.info('...saved y mean and standard deviation to {}_mean_std.npz'.format(fpath))
 
-    def predict(self, molecule, molecule_tensor=None, sigma=False):
+    def predict(self, molecule=None, molecule_tensor=None, sigma=False):
```
Comment (cgrambow) [inline cnn_framework/predictor.py]: Perhaps we should raise an exception if neither `molecule` nor `molecule_tensor` are specified.

**Review by cgrambow — CHANGES_REQUESTED** [review]
Looks good! I added a bunch of comments, but most of them just pertain to some formatting and squashing of commits.

File: cnn_framework/cnn_model.py (hunk 6 above)
Comment (yplitw) [reply cnn_framework/cnn_model.py]: This is to generate random seeds for each of the random masks, not to generate samples for bootstrapping.  So I'm not sure how we could use `np.random.choice` here.

File: cnn_framework/layers.py (hunk 1 above)
Comment (yplitw) [reply cnn_framework/layers.py:18]: The dropout mask has to be large enough to handle the largest molecule allowed in the model.  So one dimension of the mask has to be the same size as `padding_final_size`.  This is the reason why `padding_final_size` is needed in the definition of `MoleculeConv`.  If `padding_final_size=None` and dropout is used, it will crash.

File: cnn_framework/data.py (hunk 13 above)
Comment (yplitw) [reply cnn_framework/data.py:385]: I don't think resetting random seed of numpy often is a good idea (it affects globally).  A RandomState is just a local random number generator, which is more suitable here in my opinion.

File: cnn_framework/data.py (hunk 14 above)
Comment (yplitw) [reply cnn_framework/data.py]: It's not necessary.  But I feel it's better to make the code deterministic by default.  What do you think?

File: cnn_framework/uncertainty.py (hunk 5 above)
Comment (yplitw) [reply cnn_framework/uncertainty.py]: Done

File: cnn_framework/uncertainty_test.py (hunk 17 above)
Comment (yplitw) [reply cnn_framework/uncertainty_test.py]: Done

File: cnn_framework/test_data/ensemble_predictor/predictor_input.py (hunk 21 above)
Comment (yplitw) [reply cnn_framework/test_data/ensemble_predictor/predictor_input.py:1]: Done

File: cnn_framework/predictor.py (hunk 3 above)
Comment (yplitw) [reply cnn_framework/predictor.py]: Done

File: cnn_framework/predictor.py (hunk 4 above)
Comment (yplitw) [reply cnn_framework/predictor.py]: You are right. It should be `y_std`

File: cnn_framework/cnn_model_test.py (hunk 7 above)
Comment (yplitw) [reply cnn_framework/cnn_model_test.py:64]: Done

File: cnn_framework/layers.py (hunk 8 above)
Comment (yplitw) [reply cnn_framework/layers.py:10]: Done

File: cnn_framework/layers.py (hunk 10 above)
Comment (yplitw) [reply cnn_framework/layers.py:170]: Done

File: cnn_framework/molecule_tensor.py (hunk 20 above)
Comment (yplitw) [reply cnn_framework/molecule_tensor.py:205]: Done

File: cnn_framework/cnn_model.py (hunk 6 above)
Comment (cgrambow) [reply cnn_framework/cnn_model.py]: `np.random.randint` is just sampling from a uniform distribution, which you can also accomplish with `np.random.choice`. I would use something like `np.random.choice(10*n_model, n_model, replace=False)`, which would give you random seeds and ensure that none of them are the same. The only issue with `np.random.choice` is that it stores the entire uniform distribution in memory (essentially `np.arange(10*n_model)`), so you wouldn't want to use a range up to `10e8`, but that shouldn't really be an issue.

File: cnn_framework/data.py (hunk 13 above)
Comment (cgrambow) [reply cnn_framework/data.py:385]: Oh, I see, that makes sense.

File: cnn_framework/data.py (hunk 14 above)
Comment (cgrambow) [reply cnn_framework/data.py]: Yeah, perhaps deterministic is better. Some of the methods/functions in `predictor.py` and `data.py` already set the seed explicitly when calling `split_test_from_train_and_val`, etc. If we want it to be fully deterministic, we would also have to set the seed for `train_model` in `cnn_model.py` and the Keras seed to that the weights are always initialized in the same way.

File: scripts/train_cnn.py (hunk 2 above)
Comment (yplitw) [reply scripts/train_cnn.py]: Done

File: cnn_framework/cnn_model.py (hunk 6 above)
Comment (yplitw) [reply cnn_framework/cnn_model.py]: I've updated it using `range(n_model)` as we discussed.

File: cnn_framework/input_test.py (hunk 12 above)
Comment (yplitw) [reply cnn_framework/input_test.py]: Done

File: cnn_framework/molecule_tensor.py (hunk 11 above)
Comment (yplitw) [reply cnn_framework/molecule_tensor.py]: Long time ago there was no "1" in the first digit of a bond attribute.  But if we do not specify bond type, then we need this "1" there to let the model know the two atoms are bonded.  So I put a if statement there to distinguish these two cases.

File: cnn_framework/data.py (hunk 14 above)
Comment (yplitw) [reply cnn_framework/data.py]: I've added an option for users to set random seed as we discussed.

File: cnn_framework/molecule_tensor.py (hunk 15 above)
Comment (yplitw) [reply cnn_framework/molecule_tensor.py]: Done

File: cnn_framework/molecule_tensor.py (hunk 16 above)
Comment (yplitw) [reply cnn_framework/molecule_tensor.py]: Done

File: cnn_framework/predictor_test.py (hunk 18 above)
Comment (yplitw) [reply cnn_framework/predictor_test.py]: Done

File: cnn_framework/predictor_test.py (hunk 19 above)
Comment (yplitw) [reply cnn_framework/predictor_test.py]: Done

File: cnn_framework/uncertainty_test.py (hunk 22 above)
Comment (yplitw) [reply cnn_framework/uncertainty_test.py]: Done

File: cnn_framework/predictor.py (hunk 23 above)
Comment (yplitw) [reply cnn_framework/predictor.py]: Done

**Comment by cgrambow** [comment]
I pushed a quick fixup just to remove the `shuffle_seed=None` statements, since that's the default anyway. I'll go ahead and merge this.

**Review by cgrambow — APPROVED** [review]
